### PR TITLE
Firebase Integration

### PR DESCRIPTION
**Pull request by Seyi-Baks**
[Ticket](https://www.notion.so/chayn/e577e368f7374db7a01a35f32d1c6d5b?v=cfeacee3bf334b339f8d627e679a4710&p=4155b6910a89439d83b7a231ac090300) [Ticket](https://www.notion.so/chayn/e577e368f7374db7a01a35f32d1c6d5b?v=cfeacee3bf334b339f8d627e679a4710&p=d014c157b6354b60865dc7d4dec37b1e) 

This PR contains:

- Firebase Integration
- API to validate token and return user data 

## Firebase Integration 

Authentication is a crucial part of any application and Bloom - frontend/backend - uses Firebase to authenticate users. This implementation of Firebase takes a `Bearer <token>`, validates it and if valid returns the required data

## API 

`users/me ` endpoint has been created to returns user data with associated entities (Partner, PartnerAccess and PartnerAdmin as required)

### CODE REVIEW

**Review by annarhughes — COMMENTED** [review]
Looking really good overall, just a couple qs

File: src/utils/constants.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          therapySessionsRedeemed: Number(userObject.partnerAccess.therapySessionsRedeemed),
+        }
+      : {},
+    partnerAdmin: userObject.partnerAdmin
+      ? {
+          id: userObject.partnerAdmin.id,
+          createdAt: userObject.partnerAdmin.createdAt,
+          updatedAt: userObject.partnerAdmin.updatedAt,
+          userId: userObject.partnerAdmin.userId,
+          partnerId: userObject.partnerAdmin.partnerId,
```
[the remark this answers is not in the document]
Comment (Seyi-Baks) [reply src/utils/constants.ts:42]: Ideally, I wouldve liked to return the enitties as is but the way TypeOrm was returning them wasnt the best. A few nested objects here and there. This gives a little more control with how the data is structured

File: src/user/user.service.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  constructor(
+    @InjectRepository(UserRepository)
+    private userRepository: UserRepository,
+  ) {}
+
+  public async getUser({ uid }: IFirebaseUser): Promise<GetUserDto | undefined> {
+    const queryResult = await this.userRepository
+      .createQueryBuilder('user')
+      .leftJoinAndSelect('user.partnerAccess', 'partnerAccess')
+      .leftJoinAndSelect('user.partnerAdmin', 'partnerAdmin')
```
[the remark this answers is not in the document]
Comment (Seyi-Baks) [reply src/user/user.service.ts:19]: Right - I wasnt sure if a partner admin would also use that functionality but i'll remove them

File: src/interfaces/user.interface.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+export interface IUser {
+  id: string;
+  createdAt: Date | string;
+  updatedAt: Date | string;
+  firebaseUid?: string;
+  name: string;
+  email: string;
+  languageDefault: LANGUAGE_DEFAULT;
+}
```
[the remark this answers is not in the document]
Comment (Seyi-Baks) [reply src/interfaces/user.interface.ts:11]: They can go into their respective folders. Needed them when creating the getUser object

[the remark this answers is not in the document]
Comment (annarhughes) [reply src/interfaces/user.interface.ts:11]: gotcha guess they'll be needed elsewhere too if we needed it for `getUser`

File: src/utils/constants.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (annarhughes) [reply src/utils/constants.ts:42]: Good shout, this will be helpful for the frontend working with them vs lots of nesting!

**Review by annarhughes — DISMISSED** [review]

**Pull request merged**